doc/prebuilt/man/man1/corectl.1
'\" t
.\"     Title: autotune
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 2015/02/14
.\"    Manual: Xenomai Manual
.\"    Source: Xenomai 3.0.5
.\"  Language: English
.\"
.TH "AUTOTUNE" "1" "2015/02/14" "Xenomai 3\&.0\&.5" "Xenomai Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
corectl \- Cobalt core control interface
.SH "SYNOPSIS"
.sp
\fBcorectl\fR [ options ]
.SH "DESCRIPTION"
.sp
\fBcorectl\fR is a utility to
.SH "OPTIONS"
.sp
\fBcorectl\fR accepts the following options:
.PP
\fB\-\-stop [<grace\-seconds>]\fR
.RS 4
Stop the real\-time services\&. The following actions are taken in sequence:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
termination of all Xenomai threads running in user\-space, waiting for them to exit for at most
grace\-seconds
if specified, or indefinitely otherwise\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
active RTDM drivers are notified of the transition to the stopped state\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
termination of lingering RTDM threads (i\&.e\&. running in kernel space), waiting for them to exit for at most 3 seconds\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
deactivation of the real\-time timing services, control of the hardware timer on all real\-time CPUs is fully released to the host kernel\&.
.RE
.RE
.sp
Once stopped, the Cobalt core rejects all connection requests from regular applications\&.
.PP
\fB\-\-start\fR
.RS 4
Start the real\-time services\&. The following actions are taken in sequence:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
activation of the real\-time timing services, the Cobalt core takes full control over the hardware timer on all real\-time CPUs\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
loaded RTDM drivers are notified of the transition to the running state\&.
.RE
.RE
.sp
Once started, the Cobalt core accepts all connection requests from regular applications anew\&.
.PP
\fB\-\-status\fR
.RS 4
Display the current Cobalt core status\&. The following statuses can be returned:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIdisabled\fR
denotes a fully inoperative core\&. This state cannot be reached using the
\fBcorectl\fR
command, but only by passing the
xenomai\&.state=disabled
option on the kernel command line\&. A disabled core cannot be started dynamically using
\fBcorectl\fR\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIstopped\fR
means that no further connection request will be accepted from applications, the real\-time services are currently unavailable\&. The Cobalt core can be stopped at boot time by passing the
xenomai\&.state=stopped
option on the kernel command line\&. A stopped core can be started dynamically using
\fBcorectl \-\-start\fR, switching it to the
\fIrunning\fR
state\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIrunning\fR
denotes an active state of the real\-time core, application requests are processed normally\&. This is the default state entered at boot time, which corresponds to passing the
xenomai\&.state=enabled
option on the kernel command line\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIteardown\fR
denotes a real\-time system in the process of stopping all services\&. This transient status should not be seen unless some threads are unexpectedly lingering despite a termination request was issued\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\fIwarmup\fR
denotes a real\-time system in the process of starting all services\&. This transient status should not be seen unless an RTDM driver gets stuck while switching to active mode\&.
.RE
.RE
.PP
\fB\-\-help\fR
.RS 4
Display a short help\&.
.RE
.sp
If no option is passed, \-\-status is assumed by default\&.
.SH "VERSIONS"
.sp
\fBcorectl\fR appeared in Xenomai 3\&.0 for the \fICobalt\fR real\-time core\&.
.SH "AUTHOR"
.sp
\fBcorectl\fR was written by Philippe Gerum <rpm@xenomai\&.org>\&.


ChangeLog
2009-03-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vxworks/semLib.c (sem_destroy_internal) : Add missing
	rescheduling. Fix sem_read_proc().

2009-03-19  Philippe Gerum  <rpm@xenomai.org>

	* include/asm-generic/wrappers.h, src/nucleus/shadow.c: Fix
	generic bits for 2.6.29.

2009-03-18  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/pipe.c: Fix streaming mode.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/nucleus/shadow.c: Trace syscall exits.

	* patching file include/asm-generic/syscall.h,
	ksrc/nucleus/shadow.c: Add xn_sys_current_info.

	* src/rtdk/rt_print.c: suspend/resume printer thread.

	* ksrc/skins/native/syscall.c, src/skins/native/task.c: Support
	T_WARNSW on task creation.

	* src/skins/native/task.c: Avoid lazy non-TLS rt_task_self.

	* src/skins/native/task.c: Silent migration before self-deletion.

	* src/skins, ksrc/nucleus/shadow.c: Rework initial Linux
	schedparam setting.

	* ksrc/nucleus, include/asm-generic, ksrc/arch/generic: Exclude
	CPUs from Xenomai use.

	* ksrc/nucleus/shadow.c: Cleanup taskexit hook.

	* ksrc/nucleus/shadow.c: Suppress switch warning upon signal
	receipt when the process is being debugged.

	* ksrc/nucleus/sched.c: Allow to change the watchdog timeout via a
	module parameter.

	* ksrc/arch/x86/usercopy_{32, 64}.c: Fix miscompilation of
	__do_strncpy_from_user.

	* src/rtdk, include/rtdk.h: Add assert_context helpers.

2009-03-18  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/skins/posix/syscall.c: Do not return -EFAULT on zero-length
	names. Instead, return -EINVAL when trying to create objects with
	empty names.

2009-03-13  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/mx.c: Streamline mutex support.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/vrtx/task.c, src/skins/vrtx/task.c: Fix task creation.

2009-03-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/shadow.c, ksrc/nucleus/synch.c: Introduce debug
	option to detect mutexes held in relaxed sections.

2009-03-07  Philippe Gerum  <rpm@xenomai.org>

	* src/skins/posix/shm.c (__wrap_munmap): Forward to glibc upon EBADF.

	* ksrc/nucleus/Config.in, ksrc/arch/{powerpc, x86}/Config.in: Add
	missing SYS_STACKPOOLSZ switch.

2009-02-26  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/powerpc/patches: Update Adeos support to
	2.6.24.6-powerpc-DENX-2.4-03 (compatible with MPC5121-ADS
	hardware, see ksrc/arch/powerpc/patches/README).

2009-02-25  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/nmi.c: Rework NMI watchdog support.
	
2008-02-24  Stephane Fillod  <stephane.fillod@thomson.net>

	* src/skins/posix/shm.c (__wrap_mmap64): Make real call depend on
	configure detection (HAVE_MMAP64).
	
2009-02-24  Yi Li <liyi.dev@gmail.com>

	* configure.in: Remove -mfdpic switch. Set by default by
	bfin-linux-uclibc-gcc, and may cause error with bfin-uclinux-gcc.

2009-02-24  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/{arm, blackfin, powerpc}/hal.c (rthal_timer_request):
	Port x86 fix of race window.

2009-02-24  Jan Kiszka  <jan.kiszka@siemens.com>

	* ksrc/arch/x86/hal-common.c (rthal_timer_request): Fix race.

	* ksrc/nucleus/shadow.c (gatekeeper_thread): Do not refer to
	uninit sched slots.

2009-02-23  Paul Corner <paul_c@tuxcnc.org>

	* debian/prepare-patch.sh: Fix patch generation rules.

2009-02-23  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.19-x86-2.2-06,
	2.6.28.7-x86-2.2-06.

2009-02-17  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-05.

	* ksrc/arch/powerpc/patches: Upgrade to
	2.6.28.5-powerpc-DENX-2.4-04.

2009-02-13  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* src/skins/posix/shm.c (__wrap_mmap, __wrap_mmap64): Map only
	part of the underlying nucleus heap.

	* ksrc/nucleus/heap.c (xnheap_mmap): Allow partial mapping in
	user-space of a nucleus heap.

	* ksrc/nucleus/sched.c, include/nucleus/sched.h: Introduce
	sched->cpu field member.

2009-02-12  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pod.c (xnpod_start_thread): Run start hook for
	user-space threads as well.

2009-02-11  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/posix/shm.c (ftruncate): Remove GFP_USER hint.

2009-02-07  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/blackfin, include/asm-blackfin: Prepare for 2.6.28.

	* ksrc/arch/blackfin/patches: Upgrade to 2.6.28-1.9-00 (GIT).

	* ksrc/nucleus, include/nucleus: Introduce host tick deferral.

2009-02-02  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/skins/native/syscall.c (rt_queue_write): Fix memory leak.

2009-02-01  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/heap, include/asm-generic/system.h, ksrc/skins: Fix
	auto-cleanup for shared heaps.

2009-01-31  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/drivers/testing/switchtest.c,
	  src/testsuite/switchtest/switchtest.c: Add stress mode to
	switchtest.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/ia64, include/asm-ia64: Itanium port discontinued.

2009-01-30  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/arch/x86/patches: Upgrade to 2.6.27.13-x86-2.2-04,
	2.6.28.2-x86-2.2-05.

2009-01-27  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* include/asm-x86/bits/pod_{32, 64}.h (xnarch_leave_root): Fix FPU
	backup clobbering.
	
2009-01-27  Philippe Gerum  <rpm@eclipse.xenomai.org>

	* ksrc/nucleus, include/nucleus, ksrc/skins/posix, include/posix,
	sim, include/asm-sim: Introduce POSIX sporadic scheduling (PSS)
	class.

2009-01-21  Gilles Chanteperdrix  <gilles.chanteperdrix@xenomai.org>

	* ksrc/arch/arm/patches: Upgrade to 2.6.28-arm-1.2-00.
	
2009-01-19  Philippe Gerum  <rpm@xenomai.org>

	* ksrc/nucleus/pipe.c, ksrc/skins/native/pipe.c,
	ksrc/skins/rtai/fifo.c: Refactor callback management, fixing a
	severe latency issue, and a potential memory leak in both
	disconnect/release handlers.

	* ksrc/nucleus/pipe.c (xnpipe_disconnect, xnpipe_release):

utils/corectl/corectl.c
/*
 * Copyright (C) 2015 Philippe Gerum <rpm@xenomai.org>
 *
 * Xenomai is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Xenomai is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Xenomai; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 */
#include <xeno_config.h>
#include <unistd.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdio.h>
#include <error.h>
#include <sys/cobalt.h>
#include <xenomai/init.h>

int __cobalt_control_bind = 1;

static int action;

static const struct option options[] = {
	{
#define status_opt	0	/* Set this first, default action. */
		.name = "status",
		.has_arg = no_argument,
		.flag = &action,
		.val = status_opt,
	},
	{
#define stop_opt	1
		.name = "stop",
		.has_arg = optional_argument,
		.flag = &action,
		.val = stop_opt,
	},
	{
#define start_opt	2
		.name = "start",
		.has_arg = no_argument,
		.flag = &action,
		.val = start_opt,
	},
	{ /* Sentinel */ }
};

void application_usage(void)
{
        fprintf(stderr, "usage: %s <option>:\n", get_program_name());
	fprintf(stderr, "--stop [<grace-seconds>]	stop Xenomai/cobalt services\n");
	fprintf(stderr, "--start  			start Xenomai/cobalt services\n");
	fprintf(stderr, "--status			query Xenomai/cobalt status\n");
}

static int core_stop(__u32 grace_period)
{
	return cobalt_corectl(_CC_COBALT_STOP_CORE,
			      &grace_period, sizeof(grace_period));
}

static int core_start(void)
{
	return cobalt_corectl(_CC_COBALT_START_CORE, NULL, 0);
}

static int core_status(void)
{
	enum cobalt_run_states state = COBALT_STATE_DISABLED;
	int ret;

	ret = cobalt_corectl(_CC_COBALT_GET_CORE_STATUS,
			     &state, sizeof(state));
	if (ret && ret != -ENOSYS)
		return ret;

	switch (state) {
	case COBALT_STATE_RUNNING:
		printf("running\n");
		break;
	case COBALT_STATE_STOPPED:
		printf("stopped\n");
		break;
	case COBALT_STATE_DISABLED:
		printf("disabled\n");
		break;
	case COBALT_STATE_WARMUP:
		printf("warmup\n");
		break;
	case COBALT_STATE_TEARDOWN:
		printf("teardown\n");
		break;
	}

	return 0;
}

int main(int argc, char *const argv[])
{
	__u32 grace_period = 0;
	int lindex, c, ret;
	
	for (;;) {
		c = getopt_long_only(argc, argv, "", options, &lindex);
		if (c == EOF)
			break;
		if (c == '?') {
			xenomai_usage();
			return EINVAL;
		}
		if (c > 0)
			continue;

		switch (lindex) {
		case stop_opt:
			grace_period = optarg ? atoi(optarg) : 0;
		case start_opt:
		case status_opt:
			break;
		default:
			return EINVAL;
		}
	}

	switch (action) {
	case stop_opt:
		ret = core_stop(grace_period);
		break;
	case start_opt:
		ret = core_start();
		break;
	case status_opt:
		ret = core_status();
		break;
	default:
		xenomai_usage();
		exit(1);
	}

	if (ret)
		error(1, -ret, "'%s' request failed", options[action].name);

	return 0;
}
